### DISCUSSION TITLE
Unable to achieve async/parallel behavior while using Ormar

### DISCUSSION THREAD

**Discussion by kokroo** [collerek/ormar#664] [Q&A]
I am using FastAPI and have an async route that needs to do many things like making calls to other API endpoints, and reading/writing to a database. It iterates over a list of customers (around 500).

I used ThreadPoolExectuor before to achieve parallelizing of a "for loop" and reduced my execution time from 10 minutes to 5 seconds.

I was not using any "await" inside those routes and everything was fine. Now I have a database call through Omar, for example, "await Customer.objects.filter(Customer.users.contains(mac)).first()". 

This question might not be specific to Ormar, but to using the await keyword anywhere within FastAPI while trying to achieve parallelism.

Python complains and says I cannot use await in there. I Googled extensively but cannot find an answer to my problem.

Example Code:

async def customer_list():
```

 customers  =await Customer.objects.filter(Customer.users.contains(mac)).get()
 for customer in customers:
       if "Mr." in customer.name:
         //API call here
         await MaleCustomers.objects.update_or_create(**customer.dict())

```

Sequentially, waiting for the API call to complete and then the database write to finish is too slow. I would like to know if there's a way to parallelize each iteration of the for loop here?

**Comment by collerek** [comment]
You cannot reach parallel execution in this way. 

Each `await` in code returns execution to the event loop, but your code in for loop is synchronous, so there is nothing to switch to in this single request. You need to gather the tasks and await them one time only. 

That is still quite inefficient code for orm, to update multiple entities you should use `bulk_update`.

I assume that your API call is a REST API call to somewhere -> then you should parallelize this i.e. with httpx - you can check https://stackoverflow.com/questions/61537570/how-do-i-make-parallel-async-http-requests-using-httpx-versus-aiohttp-in-pytho

**Reply by kokroo** [reply]
I want to run each iteration of the loop in its own thread, and I am doing it right now. It just doesn't work when there's "await" stuff inside.

I use bulk_update elsewhere, but in this case, I need to have the DB read/write within the loop, one by one.
